Prac 09

PengPai88 opened the pull request:
@zhaofangxunjcu评论一下

PengPai88 committed:
Add a new attribute: current_fare_distance.
Add methods related to taxi fare calculation, such as get_fare() and start_fare().
The `drive()` method was rewritten to record the billing distance while driving.

PengPai88 committed:
Add a new attribute: current_fare_distance.
Add methods related to taxi fare calculation, such as get_fare() and start_fare().
The `drive()` method was rewritten to record the billing distance while driving.

PengPai88 committed:
This module defines the base Car class used for simulating vehicle behavior.
It includes attributes like fuel and odometer, and a method to drive the car.
The `drive()` method was rewritten to record the billing distance while driving.

PengPai88 committed:
"""
This module defines the SilverServiceTaxi class, a subclass of Taxi.
It adds a fanciness factor that increases the fare, and includes a flagfall charge.
"""

PengPai88 committed:
"""
Defines the UnreliableCar class that sometimes fails to drive.
It uses randomness to simulate a car that may or may not drive as intended.
"""

PengPai88 committed:
"""
This is the main program to run a Taxi fare simulation.
It lets the user choose a taxi, drive a distance, and calculates the fare.
Useful for testing Taxi and SilverServiceTaxi behavior.
"""

PengPai88 committed:
"""
This module defines a Band class that stores and manages musicians.
It can add musicians and make the whole band perform by calling each musician's play() method.
"""

PengPai88 committed:
"""
Defines the base Musician class and specific subclasses like Guitarist, Drummer, and Singer.
Each subclass can override the play() method to output a relevant message.
"""

PengPai88 committed:
"""
Creates a specific band (Extreme) using different Musician subclasses.
Demonstrates polymorphism by letting each musician do their specialized play() action.
"""

PengPai88 committed:
"""
This module defines different types of Trees:
- Tree: base class with height and growth logic.
- QuickTree: grows faster than a normal tree.
- FruitTree: bears fruit as it grows.
Useful for inheritance and method overriding examples.
"""

PengPai88 committed:
"""
Client code to test different tree types.
Creates a list of Tree, QuickTree, and FruitTree objects and simulates growth.
Prints their status before and after growing to demonstrate polymorphism.
"""

PengPai88 committed:
This code defines three subclasses of Musician:

Guitarist: plays a guitar if available, otherwise shows a message.

Drummer: always returns a message about drumming.

Singer: always returns a message about singing.

Each class overrides the play() method to suit their role.

PengPai88 committed:
This code defines the FruitTree class, a subclass of Tree:

Adds a _fruit attribute to represent the amount of fruit.

Overrides the grow() method to sometimes increase or decrease the fruit count.

Overrides get_ascii_leaves() to display fruit (.) above the tree's leaves using ASCII art.

PengPai88 committed:
This code defines the QuickTree class, a subclass of Tree:

Overrides the grow() method so the tree grows faster by directly adding water to trunk height and sunlight to number of leaves, without randomness.

PengPai88 committed:
This code snippet:

Imports the SilverServiceTaxi class.

Creates an instance fancy_taxi with name "Hummer", 200 fuel, and fanciness factor 2.

Drives the taxi for 18 km.

Prints the taxi’s details and the fare cost.

Uses an assertion to check that the fare is correctly rounded to $48.80.

PengPai88 committed:
This code defines the Guitar class which models a guitar with attributes for its name, manufacturing year, and cost. It includes:

Initialization of these attributes.

String methods (__str__ and __repr__) for readable guitar descriptions.

A method to calculate the guitar's age based on a fixed current year.

A method to determine if the guitar is vintage (older than 50 years).

A less-than method (__lt__) to enable sorting guitars by their year of manufacture.

PengPai88 committed:
This code tests the UnreliableCar class by creating a car with 30% reliability and 100 units of fuel. It attempts to drive the car 100 times for 1 km each time, accumulating the total distance actually driven (which may be less due to unreliability). Finally, it prints the total distance driven after 100 attempts.

zhaofangxunjcu commented on taxi.py:
```diff
@@ -0,0 +1,24 @@
+        distance_driven = super().drive(distance)
+        self.current_fare_distance += distance_driven
+        return distance_driven
+
```
nice naming

zhaofangxunjcu commented on prac-09/taxi_simulator.py:
```diff
@@ -0,0 +1,61 @@
+
+    while True:
+        print("q)uit, c)hoose taxi, d)rive")
+        choice = input(">>> ").lower()
```
while True is not good

PengPai88 committed:
Add files via upload

Overall Summary
These two code segments demonstrate good programming practices: 
The Single Responsibility Principle (SRP): Functions should focus on a single responsibility, such as string repetition, word length determination, and sentence formatting. 
Test-Driven Development (TDD): Use assertions and doctests to ensure the correctness of the code. 
Exception handling: By catching different types of exceptions during external API calls, the robustness of the program is enhanced. 
Third-party library usage: Make reasonable use of the Wikipedia and BeautifulSoup libraries to simplify the data scraping of Wikipedia. 
Code maintainability: Clear function annotations and docstrings facilitate understanding and expansion in the future. 
If necessary, I can help you write a commented version of the code or expand its functionality.
